src/blackmetal/src/communication/controlSW/BlackMetal.cpp

```diff
@@ -3,48 +3,65 @@
 #include "colors.hpp"
 #include "log.hpp"
 
+#include <chrono>
+#include <functional>
+
 #define PORT 665
 
 INIT_MODULE(BlackMetal, dbg_level::DBG);
 
 BlackMetal::BlackMetal()
 	: rclcpp::Node("blackmetal")
-	, Client(PORT, "192.168.1.3")
+	, m_controlClient(new Client(PORT, "192.168.1.3"))
+	, m_odometry(new Odometry(m_controlClient))
 {
-	m_chassisLength = declare_parameter<double>("chasis", 1);
-	m_wheelRadius = declare_parameter<double>("wheelRadius", 0.2);
-
-	DBG("Chasis has lenght " << m_chassisLength << " m");
-	DBG("Wheel has radius " << m_wheelRadius << " m");
-	DBG("Address set to " << address() << ":" << PORT);
-
-	m_twistSubscriber
-		= this->create_subscription<geometry_msgs::msg::Twist>(
-			"cmd_vel",
-			1,
-			[this] (const geometry_msgs::msg::Twist &msg) {
-				this->onTwistRecievedSendJson(msg);
-			}
+	m_odometry->setChassisLength(declare_parameter<double>("chasis", 1));
+	m_odometry->setWheelRadius(declare_parameter<double>("wheelRadius", 0.2));
+
+	DBG("Chasis has lenght " << m_odometry->getChassisLength() << " m");
+	DBG("Wheel has radius " << m_odometry->getWheelRadius() << " m");
+	DBG("Address set to " << m_controlClient->address() << ":" << PORT);
+
+	m_twistSubscriber = this->create_subscription<geometry_msgs::msg::Twist>(
+		"cmd_vel",
+		1,
+		[this] (const geometry_msgs::msg::Twist &msg) {
+			this->onTwistRecievedSendJson(msg);
+		}
 	);
-	// TODO: Create a service for executing the other commands.
+	WARN("Odometry created");
 }
 
 void BlackMetal::onTwistRecievedSendJson(const geometry_msgs::msg::Twist &msg)
 {
 	DBG("Message geometry_msgs::msg::Twist: " << geometry_msgs::msg::to_yaml(msg));
-	m_rightWheelSpeed = (msg.linear.x + 0.5 * m_chassisLength * msg.angular.z)/m_wheelRadius;
-	m_leftWheelSpeed = (msg.linear.x - 0.5 * m_chassisLength * msg.angular.z)/m_wheelRadius;
-	request(m_rightWheelSpeed, m_leftWheelSpeed);
+	m_rightWheelSpeed = (msg.linear.x + 0.5 * m_odometry->getChassisLength() * msg.angular.z)/ m_odometry->getWheelRadius() / 1000.;
+	m_leftWheelSpeed = (msg.linear.x - 0.5 * m_odometry->getChassisLength() * msg.angular.z)/ m_odometry->getWheelRadius() / 1000.;
+	INFO("Right wheel speed: " << m_rightWheelSpeed << " Left wheel speed: " << m_leftWheelSpeed);
+
+	checkWheelSpeed(m_rightWheelSpeed);
+	checkWheelSpeed(m_leftWheelSpeed);
+
+	m_controlClient->requestSpeed(m_rightWheelSpeed, m_leftWheelSpeed);
 }
 
-bm::Status BlackMetal::evalReturnState(const std::string &returnJson)
+const double& BlackMetal::chassisLength()
 {
-	if (returnJson.find("RECIEVE_OK") == std::string::npos) {
-		WARN("The robot buffer is full. The send data will not be used: " << returnJson);
-		return bm::Status::FULL_BUFFER;
-	}
+	return m_odometry->getChassisLength();
+}
 
-	SUCCESS(returnJson);
-	return bm::Status::OK;
+const double& BlackMetal::wheelRadius()
+{
+	return m_odometry->getWheelRadius();
+}
+
+void BlackMetal::checkWheelSpeed(double &wheelSpeed)
```

Comment:
toto robi check? A zmeni hodnotu argumentu? Ak je to check, tak by to malo vraciat bool. Ak to ma nieco modifikovat, tak by sa to malo volat modifyWheelSpeed. Alebo podla toho co to robi, asi by som to nazval clampWheelSpeed. Alebo jednoducho pouzi funkciu std::clamp :) https://en.cppreference.com/w/cpp/algorithm/clamp

Reply:
jo opravene